src/utils/deviceInfo/deepDeviceInfo.ts

```diff
@@ -0,0 +1,172 @@
+import UAParser from 'ua-parser-js';
+import { DetailedDeviceInformation, SupportedPlatforms } from '../../types/device';
+
+interface DetailedDeviceInformationParserProps {
+    deviceInformation: DetailedDeviceInformation;
+    uaDataValues?: UADataValues;
+    uaParser: UAParser;
+}
+
+/**
+ * This object is concerned with gathering and holding information about the users device.
+ * This object will be read-only after initialization and should as a rule throw errors if the
+ * deviceInformation property is altered after intialization.
+ */
+class DetailedDeviceInformationProvider {
+    /**
+     * Represents the user agent string parser.
+     * This method is treated as a fallback if the User-Agent Client Hints API does not provide a coherent value
+     * or if the browser does not yet support it.
+     */
+    private readonly _uaParser: UAParser;
+
+    /**
+     * Stores the system report from a call to the underlying system.
+     * More info: https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API
+     */
+    private readonly _uaDataValues: UADataValues | undefined = undefined;
+
+    /* * Holds the current device information and should be the single source of truth. */
+    private readonly _deviceInformation: DetailedDeviceInformation;
+
+    private constructor(props: DetailedDeviceInformationParserProps) {
+        this._deviceInformation = props.deviceInformation;
+        this._uaDataValues = props.uaDataValues;
+        this._uaParser = props.uaParser;
+    }
+
+    public get deviceInformation(): DetailedDeviceInformation {
+        return this._deviceInformation;
+    }
+
+    public get uaDataValues() {
+        return this._uaDataValues;
+    }
+
+    public get uaParser() {
+        return this._uaParser;
+    }
+
+    /** Constructs the parser and returns it.  */
+    public static initialize(): DetailedDeviceInformationProvider {
+        // Assign essentially an alias for readability.
+        const DDIP = DetailedDeviceInformationProvider;
+        const navigatorUAData: NavigatorUAData | undefined = navigator?.userAgentData;
+        const uaParser = new UAParser(navigator.userAgent);
+        let deviceInformation: DetailedDeviceInformation | undefined;
+
+        // User-Agent Client Hints API is not supported, use parser exclusively instead.
+        if (!navigatorUAData) {
```

Comment:
could the difference between User-Agent Client Hints API being supported and not be split into two functions?
Something like functionnames `useUAParserExclusively` and `useClientHintsWithUAParserFallback`